### PR TITLE
Fragmentation mechanism

### PR DESCRIPTION
**Pull request by thvdveld**
This contains the fragmentation mechanism used in #580 (which is heavily based on #236 taking into account the remarks that were made there). I tried to decouple it from the 6LoWPAN implementation such that it can be used for other protocols as well.

I opened a separate PR for this such that #580 will only contain the mechanisms that are part of 6LoWPAN only.

### CODE REVIEW

File: src/storage/assembler.rs [hunk 1]
```diff
@@ -184,8 +193,9 @@ impl Assembler {
 
     /// Add a new contiguous range to the assembler, and return `Ok(())`,
     /// or return `Err(())` if too many discontiguities are already recorded.
-    pub fn add(&mut self, mut offset: usize, mut size: usize) -> Result<(), TooManyHolesError> {
+    pub fn add(&mut self, mut offset: usize, mut size: usize) -> Result<bool, TooManyHolesError> {
```
Comment (aramperes) [inline src/storage/assembler.rs]: Doc comment should be updated to reflect signature change (the `return Ok(())` bit)

Comment (thvdveld) [reply src/storage/assembler.rs]: Updated now, thanks!

**Comment by thvdveld** [comment]
Friendly reminder: could someone review this PR? @Dirbaio @aramperes

File: src/iface/fragmentation.rs [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+use managed::{ManagedMap, ManagedSlice};
+
+use crate::storage::Assembler;
+use crate::time::Instant;
+use crate::Error;
+use crate::Result;
+
+pub trait PacketAssemblerInfo: PartialEq {
+    /// Calculate a new offset based on some other information.
+    fn calc_offset(&self, offset: usize) -> usize;
```
Comment (Dirbaio) [inline src/iface/fragmentation.rs]: Why is this needed? from the 6lowpan code it seems it adds a fixed offset, could the calling code do that on its own instead?

File: src/iface/fragmentation.rs [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+
+                let len = data.len();
+                self.buffer[offset..][..len].copy_from_slice(data);
+
+                match assembler.add(offset, data.len()) {
+                    Ok(false) => {
+                        *last_updated = now;
+                        self.is_complete()
+                    }
+                    Ok(true) => Err(Error::PacketAssemblerOverlap),
```
Comment (Dirbaio) [inline src/iface/fragmentation.rs]: is it required to drop packets with overlap? AFAIK it's not for IPv4 fragmentation. With that, we could avoid making changes to `Assembler`

File: src/iface/fragmentation.rs [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+                if self.is_complete()? {
+                    let a = &self.buffer[..total_size];
+                    self.assembler = AssemblerState::NotInit;
+                    a
+                } else {
+                    return Err(Error::PacketAssemblerIncomplete);
+                }
+            }
+        };
+        self.assembler = AssemblerState::NotInit;
```
Comment (Dirbaio) [inline src/iface/fragmentation.rs]: this line is duplicated

File: src/iface/fragmentation.rs [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+    ///
+    /// - Returns [`Error::PacketAssemblerSetFull`] when every [`PacketAssembler`] in the buffer is used (only
+    /// when the non allocating version of is used).
+    pub(crate) fn reserve_with_key(&mut self, key: &K) -> Result<&mut PacketAssembler<'a, Info>> {
+        if self.packet_buffer.len() == self.index_buffer.len() {
+            match &mut self.packet_buffer {
+                ManagedSlice::Borrowed(_) => return Err(Error::PacketAssemblerSetFull),
+                #[cfg(any(feature = "std", feature = "alloc"))]
+                ManagedSlice::Owned(b) => {
+                    b.resize_with(self.index_buffer.len() + 1, || {
```
Comment (Dirbaio) [inline src/iface/fragmentation.rs]: this could be a .push(), we're just pushing 1 elem.

File: src/iface/fragmentation.rs [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+                }
+            }
+        }
+
+        let i = self
+            .get_free_packet_assembler()
+            .ok_or(Error::PacketAssemblerSetFull)?;
+
+        // NOTE(thvdveld): this should not fail because we already checked the available space.
+        match self.index_buffer.insert(*key, i as u8) {
```
Comment (Dirbaio) [inline src/iface/fragmentation.rs]: This can overflow the u8 when using `alloc` (or if the user gave space for more than 256 slots) and someone spams us with lots of incomplete fragmented packets.

I think we should have a cap on WIP reassembling packets, even when using alloc, otherwise we're vulnerabile to memory exhaustion attacks if someone spams us.

File: src/iface/fragmentation.rs (hunk 3 above)
Comment (thvdveld) [reply src/iface/fragmentation.rs]: Indeed, you are right, I misread the RFC. When there is an overlap AND either the size or the offset is different of the overlapping packet, the packet needs to be dropped in the assembler and a fresh reassembly may start with the new packet. However, I currently don't save the size of the packet in the assembler, so there is no way of knowing that. I guess for now I can remove the dropping.

File: src/iface/fragmentation.rs (hunk 5 above)
Comment (thvdveld) [reply src/iface/fragmentation.rs]: Changed it.

File: src/iface/fragmentation.rs (hunk 6 above)
Comment (thvdveld) [reply src/iface/fragmentation.rs]: I currently set the cap to u8::MAX and change the index to usize instead of u8. This should prevent overflow and cap the WIP reassembling packets.

File: src/iface/fragmentation.rs (hunk 2 above)
Comment (thvdveld) [reply src/iface/fragmentation.rs]: Maybe my solution is not the best. Here is what I tried to solve:

The size in the first fragment header represents the total size after decompression. However, the packets that arrive are compressed and I put the compressed packets in the fragmentation buffers. This means that I need to calculate the total size of the compressed form, which is the total size from the first fragment header, minus the uncompressed size (of all headers that are compressed), plus the compressed size (of all headers that are compressed). Also the offset in the fragment headers are the offset in the uncompressed form. Which also means that I need to recalculate the offset to the compressed format. I have no idea why they made it so difficult (probably to prevent some attacks, though I'm not sure).

I could try to do the following in 6LoWPAN to get rid of `PacketAssemblerInfo`: just decompress the packet before putting it in the packet assembler buffers. I didn't do this in the first place because then I would decompress in two places: first when we there is no fragmentation and second for fragmented packets.

File: src/iface/fragmentation.rs (hunk 3 above)
Comment (thvdveld) [reply src/iface/fragmentation.rs]: Changed it.

File: src/iface/fragmentation.rs (hunk 2 above)
Comment (thvdveld) [reply src/iface/fragmentation.rs]: I removed it for now in this PR and will see later how I will fix it for 6LoWPAN.

**Review by Dirbaio — APPROVED** [review]
Thank you!

Sorry for the mega late reply, I've been swamped with work :( 

bors r+
